File: problems-extend/0801-0900/0865_smallest-subtree-with-all-the-deepest-nodes/solution.ts
function subtreeWithAllDeepest(root: TreeNode | null): TreeNode | null {
    if (root === null) {
        return null;
    }
    // A node can only be judged once both of its children's heights are
    // known, so the walk is post-order — children before the node — on
    // an explicit stack of {node, measured} frames: the first pop pushes
    // the node's own merge beneath its two children, and that merge —
    // the second pop — can only fire once both subtrees are measured.
    // Iterating keeps a 500-node chain's ~500 merges off the 512k V8
    // stack this judge runs Node with.
    const heights = new Map<TreeNode, number>();
    const smallest = new Map<TreeNode, TreeNode>();
    const stack: Array<{ node: TreeNode; measured: boolean }> = [{ node: root, measured: false }];
    while (stack.length > 0) {
        const frame = stack.pop()!;
        if (!frame.measured) {
            stack.push({ node: frame.node, measured: true });
            if (frame.node.right !== null) {
                stack.push({ node: frame.node.right, measured: false });
            }
            if (frame.node.left !== null) {
                stack.push({ node: frame.node.left, measured: false });
            }
            continue;
        }
        const lh = frame.node.left === null ? 0 : heights.get(frame.node.left)!;
        const rh = frame.node.right === null ? 0 : heights.get(frame.node.right)!;
        heights.set(frame.node, 1 + Math.max(lh, rh));
        // Equal heights: each side reaches this subtree's deepest level,
        // so its deepest nodes sit on both sides and only this node
        // covers them all — it is the subtree's answer. Unequal: no
        // deepest node can live in the shallower side, so the deeper
        // side's answer passes through unchanged.
        if (lh === rh) {
            smallest.set(frame.node, frame.node);
        } else {
            const deeper = lh > rh ? frame.node.left! : frame.node.right!;
            smallest.set(frame.node, smallest.get(deeper)!);
        }
    }
    return smallest.get(root)!;
}

File: problems-extend/0801-0900/0865_smallest-subtree-with-all-the-deepest-nodes/solution.go
// frame pairs a subtree root with whether its height is measured yet.
type frame struct {
	node     *TreeNode
	measured bool
}

func subtreeWithAllDeepest(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	// A node can only be judged once both of its children's heights are
	// known, so the walk is post-order — children before the node — over
	// an explicit stack: the first pop of a frame schedules the node's own
	// merge beneath its two children, and the second — the measured frame
	// — can only fire once both subtrees are measured. Iterating keeps a
	// 500-node chain's ~500 merges off the goroutine stack.
	heights := map[*TreeNode]int{}
	smallest := map[*TreeNode]*TreeNode{}
	stack := []frame{{root, false}}
	for len(stack) > 0 {
		top := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if !top.measured {
			stack = append(stack, frame{top.node, true})
			if top.node.Right != nil {
				stack = append(stack, frame{top.node.Right, false})
			}
			if top.node.Left != nil {
				stack = append(stack, frame{top.node.Left, false})
			}
			continue
		}
		node := top.node
		lh, rh := 0, 0
		if node.Left != nil {
			lh = heights[node.Left]
		}
		if node.Right != nil {
			rh = heights[node.Right]
		}
		heights[node] = 1 + max(lh, rh)
		// Equal heights: each side reaches this subtree's deepest level,
		// so its deepest nodes sit on both sides and only this node
		// covers them all — it is the subtree's answer. Unequal: no
		// deepest node can live in the shallower side, so the deeper
		// side's answer passes through unchanged.
		if lh == rh {
			smallest[node] = node
		} else if lh > rh {
			smallest[node] = smallest[node.Left]
		} else {
			smallest[node] = smallest[node.Right]
		}
	}
	return smallest[root]
}
